FormalSystems/Computability/Loop.lean

```diff
@@ -22,18 +22,52 @@ inductive LoopProgram
 | loop (x: Variable) (p: LoopProgram)
 deriving DecidableEq
 
-def LoopProgram.toFunction (p: LoopProgram):
-  Nat → Nat :=
-  -- Todo: define how to evaluate `p`, with x₀ bound to the input
-  -- for example one could use a state monad with an `AssocList`
-  sorry
-
 def LoopProgram.len: LoopProgram → Nat
 | add x y n => x.index + y.index + n + 1
 | sub x y n => x.index + y.index + n + 1
 | concat p₁ p₂ => p₁.len + p₂.len + 1
 | loop n p => p.len + n.index + 1
 
+def LoopProgramState := Lean.AssocList Variable Nat 
+
+namespace LoopProgramState
+  def get (s: LoopProgramState) (v: Variable): Nat := (s.find? v).getD 0
+
+  def put (s: LoopProgramState) (v: Variable) (n: Nat): LoopProgramState := 
+    if s.contains v then s.replace v n else s.insert v n
+end LoopProgramState
+
+mutual
+  def LoopProgram.runNTimes (p: LoopProgram) (n: Nat) (s: LoopProgramState): LoopProgramState :=
+    match n with 
+      | .zero => s 
+      | .succ m => p.runNTimes m (p.run s)
+
+  def LoopProgram.run: LoopProgram -> LoopProgramState -> LoopProgramState
+  | add x y n => fun s => 
+    let new_val := s.get y + n
+    s.put x new_val
+  | sub x y n => fun s =>
+    let new_val := s.get y - n
+    s.put x new_val
+  | concat p q => fun s => 
+    have : len p < len (concat p q) := by simp [len]; apply Nat.lt_of_succ_le; apply Nat.succ_le_succ; simp
+    have : len q < len (concat p q) := by simp [len]; apply Nat.lt_of_succ_le; apply Nat.succ_le_succ; simp
+    q.run (p.run s)
+  | loop x p => fun s => 
+    have : len p < len (loop x p) := by simp [len]; apply Nat.lt_of_succ_le; apply Nat.succ_le_succ; simp
+    p.runNTimes (s.get x) s
```

Comment:
```suggestion
  def LoopProgram.run: LoopProgram -> LoopProgramState -> LoopProgramState
  | add x y n => fun s =>  
    let new_val := s.get y + n
    s.put x new_val
  | sub x y n => fun s =>
    let new_val := s.get y - n
    s.put x new_val
  | concat p q => fun s => 
    q.run (p.run s)
  | loop x p => fun s => 
    Nat.iterate p.run (s.get x) s
```
When no mutual block is used, lean is actually able to check termination by itself in this case.

Reply:
I find it kind of curious that the termination check just works here but I take it.
It still kind of makes sense to me because `p.run` is fixed here when calling `Nat.iterate` whereas with the mutual approach only `p` was passed.